### ISSUE TITLE
Int8 Inference Slowdown Comparing to FP32 when using PyTorch 2 Export Quantization with X86 Backend through Inductor

### ISSUE THREAD

**Issue by YixuanSeanZhou** [pytorch/pytorch#144434]
### 🐛 Describe the bug

INT8 inference is significantly slower comparing to FP32 after performing quantization using the tutorial: [PyTorch 2 Export Quantization with X86 Backend through Inductor](https://pytorch.org/tutorials/prototype/pt2e_quant_x86_inductor.html)

When benchmarking the inference speed I observe: 
```
Python before quant inference time: 0.022294288396835327

Python inference time after quant: 1.5116438150405884
```

A complete repro of the issue: 
```
import torch
import torchvision
import tempfile
import time
import sys
import os

torch.manual_seed(42)

from torch._export import capture_pre_autograd_graph
from torch.ao.quantization.quantize_pt2e import convert_pt2e, prepare_pt2e
# from torch.ao.quantization.quantizer.xnnpack_quantizer import (
#     XNNPACKQuantizer, get_symmetric_quantization_config,
# )

from torch.ao.quantization.quantizer.x86_inductor_quantizer import (
    X86InductorQuantizer, get_default_x86_inductor_quantization_config
)

import torch._inductor.config as config
config.cpp_wrapper = True

RUN_NUM = 1000

def print_size_of_model(model):
    torch.save(model.state_dict(), "temp.p")
    print("Size (MB):", os.path.getsize("temp.p")/1e6)
    os.remove("temp.p")

def main(argv):
    workdir="/tmp"

    # Step 1: Capture the model
    args = (torch.randn(1, 3, 224, 224),)
    m = torchvision.models.resnet18(weights=torchvision.models.ResNet18_Weights.DEFAULT).eval()
    m1 = m.eval()
    

    print("before size")
    print_size_of_model(m)


    with torch.no_grad():
        compiled = torch.compile(m1, backend="inductor")
        before = time.time()
        for i in range(RUN_NUM):
            compiled(*args)
        print(f"Python inference time: {(time.time() - before) / RUN_NUM}")


    # torch.export.export_for_training is only avaliable for torch 2.5+
    m = capture_pre_autograd_graph(m, args)

    # Step 2: Insert observers or fake quantize modules
    # quantizer = XNNPACKQuantizer().set_global(
    #     get_symmetric_quantization_config())
    quantizer = X86InductorQuantizer().set_global(
        get_default_x86_inductor_quantization_config())
    m = prepare_pt2e(m, quantizer)

    # Step 2.5 fake calibration
    m(*args)
    # # Step 3: Quantize the model
    m = convert_pt2e(m, fold_quantize=True)
    
    print(m)
    print("after size")
    print_size_of_model(m)


    with torch.autocast(device_type="cpu", dtype=torch.bfloat16, enabled=True), torch.no_grad():
        before_compile_time = time.time()
        compiled_quant = torch.compile(m, backend="inductor")
        print("finnished compiliation", time.time() - before_compile_time)
        before = time.time()
        for i in range(RUN_NUM // 100):
            compiled_quant(*args)
            # print("finished run", i)
        print(f"Python inference before quant time after quant: {(time.time() - before) / (RUN_NUM // 100)}")


    os.makedirs(workdir, exist_ok=True)

if __name__ == '__main__':
    main(sys.argv)
```

### Versions

Versions
```
python collect_env.py
Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.1.105
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] torch==2.5.1+cu121
[pip3] torchaudio==2.5.1+cu121
[pip3] torchvision==0.20.1+cu121
[pip3] triton==3.1.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.1.105                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] torch                     2.5.1+cu121              pypi_0    pypi
[conda] torchaudio                2.5.1+cu121              pypi_0    pypi
[conda] torchvision               0.20.1+cu121             pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @msaroufim

**Comment by leslie-fang-intel** [comment]
@Xia-Weiwen Could you help to take a look?

**Comment by YixuanSeanZhou** [comment]
@Xia-Weiwen Thanks for the pointers, after doing 5 warm-up run and TORCHINDUCTOR_FREEZING=1 I am getting
```
Python inference time before quant: 0.009380168437957763
Python inference time after quant: 0.007475273370742798
```
Is this expected? I would assuming going from FP32 -> INT8 should give ~2x - 4x speedup comparing to the original model? 
(When experimenting resnet with TRT + GPU, I was able to see a close to 2x speedup)

**Comment by Xia-Weiwen** [comment]
@YixuanSeanZhou I tried your code and performance looks pretty good on a 4th gen Intel Xeon CPU. Since your script compares performance between no-quant + eager and quant + compile, the speedup is quite significant (>2x). You may do the following things:
- Collect CPU info for analysis
- Run profiling with `torch.profiler` to find bottleneck.
- Run your script with DNNL_VERBOSE=1 to collect oneDNN logs.
- Bind cores when running your script by `numactl -C <core list>` although it is unlikely to be an issue to get speedup.

**Comment by YixuanSeanZhou** [comment]
Thank you @Xia-Weiwen for your investigation and help! 

> between no-quant + eager

I thought i am comparing a compiled non-quantzied model? Here is the code where i was running the compiled non-quantized model in the repro code above:
```
        compiled = torch.compile(m1, backend="inductor")
        # Added warmup later
        before = time.time()
        for i in range(RUN_NUM):
            compiled(*args)
        print(f"Python inference time: {(time.time() - before) / RUN_NUM}")
```
Are you able to see the speedup even with the non-quantized model being compiled? 

Thank you for providing the steps how to analyze the bottleneck, I will take a look at those later!

**Comment by Xia-Weiwen** [comment]
Oh. I missed that in your script. If I also run warm-up for non-quantization, the speedup ratio is not that high. And I found that you will get different speedup ratio with different core-bindings.

**Comment by Xia-Weiwen** [comment]
Hi @YixuanSeanZhou Any updates on this? Thanks.

**Comment by YixuanSeanZhou** [comment]
Hey @Xia-Weiwen, here is what I have so far

First, `numactl -C <core list>` this didn't really help me achieve more speed up.

Then, I tried `DNNL_VERBOSE=1` and seeing logs like
`onednn_verbose,primitive,exec,cpu,convolution,jit_uni_int8:avx2,forward_inference,src_u8::blocked:acdb::f0 wei_s8:a:blocked:ABcd2b8a4b::f0 bia_f32:a:blocked:a::f0 dst_u8::blocked:acdb::f0,attr-scratchpad:user attr-scales:src0:0:f32+dst:0:f32+wei:1:f32 attr-post-ops:sum:0.00446517+eltwise_relu,alg:convolution_direct,mb1_ic512oc512_ih7oh7kh3sh1dh0ph1_iw7ow7kw3sw1dw0pw1,0.35791` after compile. So I believe the INT8 kernels are being used. 

So I looked at the profile, for the int8 inference (see below), it appears the final linear layer is very slow. 
![Image](https://github.com/user-attachments/assets/6320f94c-cb29-4bc8-8cd4-5f727fb5bd35)

I tried to skip quantizing that layer by doing: 
```
quantizer = X86InductorQuantizer().set_global(
                get_default_x86_inductor_quantization_config())
quantizer.set_module_type_qconfig(torch.nn.Linear, None)
```

This gives me slightly better result than before, but still not comparable with GPU + TRT: 
```
time before quant: 0.00987846851348877
time after quant: 0.006491565704345703 (before was 0.0074) 
```
![Image](https://github.com/user-attachments/assets/d679b08c-9ccd-41e2-9cd6-c0139fd0ceae)


I guess my question here is: is this the expected result for CPU speed up? I understand that Nvidia GPU has tensor core that can only be utilized by INT8 / FP16 so this will naturally give us speed up when lowered from FP32. But I guess CPU doesn't have such structure? So would this result be expected?

**Comment by YixuanSeanZhou** [comment]
Furthermore, do you think whether it is possible for PyTorch team to provide a table of such benchmarks? What I am thinking is a table of the following attribute: 
- Model Structure
- Quantization Method
- Harware (maybe one GPU, one CPU)
- FP32 Speed
- INT8 Speed

I think this would greatly benefits POCs and help PyTorch users set clear expectations when trying torch quantization.

**Comment by Xia-Weiwen** [comment]
Hi @YixuanSeanZhou We have some benchmark data here: https://dev-discuss.pytorch.org/t/torchinductor-update-6-cpu-backend-performance-update-and-new-features-in-pytorch-2-1/1514
As for your case, I guess it's because your CPU is old or it's an AMD CPU, which only supports AVX2. The kernel is not optimal on your platform or it has a bug. Can you share your CPU info from `lscpu`?
From your log, the last layer of linear was very slow. It is not expected. Can you share the oneDNN verbose log of that layer for both quantized and non-quantized case? You may just copy the log of the last iteration. And I guess you are using oneDNN v3.5.3, which is integrated in PyTorch 2.5. You may find the version number of oneDNN at the beginning of logs. Please confirm that.

**Comment by YixuanSeanZhou** [comment]
@Xia-Weiwen, Thanks! The shared discussion is very useful! I didn't know this resources. I will sign up and follow along. 

Yes my CPU is pretty old I think and it is not an AMD CPU, I think this could be the cause! But I am still surprised that linear takes that long Here is the output: 
```
$ lscpu
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 48 bits virtual
CPU(s):                               16
On-line CPU(s) list:                  0-15
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                63
Model name:                           Intel(R) Xeon(R) CPU @ 2.30GHz
Stepping:                             0
CPU MHz:                              2299.998
BogoMIPS:                             4599.99
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            256 KiB
L1i cache:                            256 KiB
L2 cache:                             2 MiB
L3 cache:                             45 MiB
NUMA node0 CPU(s):                    0-15
```

My oneDNN is also v3.5.3 as you expected
```
onednn_verbose,info,oneDNN v3.5.3 (commit 66f0cb9eb66affd2da3bf5f8d897376f04aae6af)
```

**Comment by Xia-Weiwen** [comment]
@YixuanSeanZhou Thanks for you info. But I didn't find any info about linear in your oneDNN log. Can you please check if you have missed anything? Please remove this `quantizer.set_module_type_qconfig(torch.nn.Linear, None)`

**Comment by Xia-Weiwen** [comment]
I see. From the log you may see it went into the reference kernel which is super slow:
- onednn_verbose,primitive,exec,cpu,matmul,**ref:any**,undef,src_u8:a:blocked:ab::f0 wei_s8::blocked:ab::f0 bia_f32:a:blocked:ab::f0_mask2 dst_bf16:a:blocked:ab::f0,attr-scratchpad:user attr-scales:src0:0:f32+wei:2:f32,,1x512:512x1000,3.62183

So, it seems that oneDNN did not optimize int8 matmul on machines with AVX2. Are you able to use a newer machine? If not, you may try FX quantization although it is deprecated: https://pytorch.org/docs/stable/quantization.html#prototype-maintenance-mode-fx-graph-mode-quantization Please use `x86` backend instead of the `qnnpack` backend in the examples.

**Comment by YixuanSeanZhou** [comment]
Got it, @Xia-Weiwen , thank you so much for your help. I will give FX graph a try. 

It won't be easy for me to upgrade system right now but maybe in the near future my team will consider upgrading to AMD / Qualcomm CPU based chips. Right now, the effort is exploratory and mostly trying to set some baselines around GPU Quantization (what we actively work on). 

I will close this ticket for now but might bug you for help future. Might open more in the future when I get there.

Thank you!

**Comment by Xia-Weiwen** [comment]
If you want to try a newer Intel CPU, you may get an instance on AWS or other cloud service providers. The cost won't be high I think. Please note that many optimizations for CPU in PyTorch are only implemented and validated for Intel CPUs although they may also benefit AMD CPUs as they are all x86 CPUs. And it would be another story for ARM based CPUs. You may try them out.
Please go ahead to close this ticket if it is ok. Thanks.

**Comment by Xia-Weiwen** [comment]
BTW, you may also try https://github.com/intel/intel-extension-for-pytorch for quantization and more features on x86 CPU if you want to.

**Comment by YixuanSeanZhou** [comment]
Got it, sry I forgot to hit close. 

One more QQ: is it fair to say that Torch is mostly focused on Intel CPU quantization and not ARM CPU?

**Comment by Xia-Weiwen** [comment]
> One more QQ: is it fair to say that Torch is mostly focused on Intel CPU quantization and not ARM CPU?

No. I didn't mean that. I meant ARM has its own implementation and optimization. They might be different from x86. And as far as I know, many features are enabled on x86 CPUs earlier than on ARM CPUs. So, don't be surprised if you find something that is only available on x86 CPUs not on ARM CPUs.
